rs/nervous_system/agent/src/helpers/nns.rs

```diff
@@ -156,3 +163,54 @@ pub async fn convert_icp_to_cycles<C: CallCanisters>(benecificary_agent: &C, amo
         .unwrap()
         .unwrap();
 }
+
+pub async fn add_wasm_via_nns_proposal<
+    C: CallCanistersWithStoppedCanisterError + ProgressNetwork,
+>(
+    agent: &C,
+    neuron_id: NeuronId,
+    wasm: SnsWasm,
+) -> Result<ProposalInfo, String> {
+    let hash = wasm.sha256_hash();
+    let canister_type = wasm.canister_type;
+    let payload = AddWasmRequest {
+        hash: hash.to_vec(),
+        wasm: Some(wasm),
+    };
+
+    let proposal = MakeProposalRequest {
+        title: Some(format!("Add WASM for SNS canister type {}", canister_type)),
+        summary: "summary".to_string(),
+        url: "".to_string(),
+        action: Some(ProposalActionRequest::ExecuteNnsFunction(
+            ExecuteNnsFunction {
+                nns_function: NnsFunction::AddSnsWasm as i32,
+                payload: Encode!(&payload).expect("Error encoding proposal payload"),
+            },
+        )),
+    };
+    propose_and_wait(agent, neuron_id, proposal).await
+}
+
+pub async fn get_nns_neuron_controller<C: CallCanisters>(
+    agent: &C,
+    neuron_id: NeuronId,
+) -> Result<Option<PrincipalId>, String> {
+    let request = ListNeurons {
+        neuron_ids: vec![neuron_id.id],
+        ..Default::default()
+    };
+    let response = list_neurons(agent, request)
```

Comment:
So this will only work if the agent represents a caller that is a hotkey, controller, or the requested neuron follows on the ManageNeuron topic.  

So that may really limit the utility of this method in tests

Reply:
That's true, but I guess that's the expected behaviour of this canister method. `get_nns_neuron_controller` is not considered to be test-only. In tests, we usually know hotkeys/controller in advance, so this shouldn't introduce any issues, while actual users (outside of tests) should get more-or-less user-friendly error messages when trying to get the controller of the neuron without sufficient rights